File: tools/java/crcl4java/examples/javascript/rhino/crclclient.js


// Import all the Java types we will need.
var urls = java.lang.reflect.Array.newInstance(java.net.URL, 1);
urls[0] = new java.net.URL("file:crcl4java-utils-1.3-jar-with-dependencies.jar");
var classloader = new java.net.URLClassLoader(urls);
var CRCLSocket = classloader.loadClass("crcl.utils.CRCLSocket");
var CRCLPosemath = classloader.loadClass("crcl.utils.CRCLPosemath");
var CRCLCommandInstanceType = classloader.loadClass("crcl.base.CRCLCommandInstanceType");
var InitCanonType = classloader.loadClass("crcl.base.InitCanonType");
var MoveToType = classloader.loadClass("crcl.base.MoveToType");
var GetStatusType = classloader.loadClass("crcl.base.GetStatusType");
var PoseType = classloader.loadClass("crcl.base.PoseType");
var PointType = classloader.loadClass("crcl.base.PointType");
var VectorType = classloader.loadClass("crcl.base.VectorType");
var BigInteger = java.math.BigInteger;

function pose(pt,xaxis,zaxis) {
    var pose = PoseType.newInstance();
    pose.setPoint(pt);
    pose.setXAxis(xaxis);
    pose.setZAxis(zaxis);
    return pose;
}

function point(x,y,z) {
    var pt = PointType.newInstance();
    pt.setX(java.math.BigDecimal.valueOf(x));
    pt.setY(java.math.BigDecimal.valueOf(y));
    pt.setZ(java.math.BigDecimal.valueOf(z));
    return pt;
}

function vector(i,j,k) {
    var v= VectorType.newInstance();
    v.setI(java.math.BigDecimal.valueOf(i));
    v.setJ(java.math.BigDecimal.valueOf(j));
    v.setK(java.math.BigDecimal.valueOf(k));
    return v;
}

var parmtypes = java.lang.reflect.Array.newInstance(java.lang.Class, 1);
parmtypes[0]=java.net.Socket;
var constargs = java.lang.reflect.Array.newInstance(java.lang.Object, 1);
constargs[0]=new java.net.Socket("localhost", 64444);
var socket = CRCLSocket.getConstructor(parmtypes).newInstance(constargs);
// Create an instance to wrap all commands.
print("s=" + socket);
var instance = CRCLCommandInstanceType.newInstance();
print("instance=" + instance);

var init = InitCanonType.newInstance();
init.setCommandID(BigInteger.valueOf(7));
instance.setCRCLCommand(init);
socket.writeCommand(instance);

// Create and send MoveTo command.
var moveTo = MoveToType.newInstance();
moveTo.setCommandID(BigInteger.valueOf(8));
var pose = pose(point(0.65,0.05,0.15),vector(1,0,0),vector(0,0,1));
moveTo.setEndPosition(pose);
moveTo.setMoveStraight(false);
instance.setCRCLCommand(moveTo);
socket.writeCommand(instance);
print("moveTo="+moveTo);

// Create and send getStatus request.
var getStat = GetStatusType.newInstance();
getStat.setCommandID(BigInteger.valueOf(9));
instance.setCRCLCommand(getStat);
socket.writeCommand(instance);

// Read status from server
var stat = socket.readStatus();
print("stat="+stat);

// Print out the status details.
var cmdStat = stat.getCommandStatus();
var IDback = cmdStat.getCommandID();
print("Status:");
print("CommandID = " + IDback);
print("State = " + cmdStat.getCommandState());
pt = stat.getPoseStatus().getPose().getPoint();
print("pt = " + pt.getX() + "," + pt.getY() + "," + pt.getZ());
var jst = stat.getJointStatuses();
if(jst != null) {
    var l = jst.getJointStatus();
    print("Joints:");
    for (index=0; index<l.size(); index++) {
        var js = l.get(index);
        print("Num=" + js.getJointNumber() + " Pos=" + js.getJointPosition());
    }
}

print("Closing socket");
socket.close();

File: tools/java/crcl4java/examples/C++/crclClient.cpp

#include "crclj.h"
#include <iostream>
#include <stdlib.h>

using namespace crclj;
using namespace crclj::crcl::base;
using namespace crclj::crcl::utils;
using namespace crclj::java::math;
using namespace crclj::java::util;

using namespace std;

int main(int argc, const char **argv) {

    try {
        // Connect to the server
        CRCLSocket s("localhost", 64444);
        
        // Create an instance to wrap all commands.
        CRCLCommandInstanceType instance;

        // Create and send init command.
        InitCanonType initCmd;
        initCmd.setCommandID(BigInteger::valueOf(7));
        instance.setCRCLCommand(initCmd);
        s.writeCommand(instance);

        // Create and send MoveTo command.
        MoveToType moveTo;
        moveTo.setCommandID(BigInteger::valueOf(8));
        PoseType pose;
        PointType pt;
        pt.setX(BigDecimal::valueOf(0.6));
        pt.setY(BigDecimal::valueOf(0.1));
        pt.setZ(BigDecimal::valueOf(0.1));
        PrintObject("pt.getX()=",pt.getX());
        PrintObject("BigDecimal::valueOf(0.6)=",BigDecimal::valueOf(0.6));
        pose.setPoint(pt);
        VectorType xAxis;
        xAxis.setI(BigDecimal::getONE());
        xAxis.setJ(BigDecimal::getZERO());
        xAxis.setK(BigDecimal::getZERO());
        pose.setXAxis(xAxis);
        VectorType zAxis;
        zAxis.setI(BigDecimal::getZERO());
        zAxis.setJ(BigDecimal::getZERO());
        zAxis.setK(BigDecimal::getONE());
        pose.setZAxis(zAxis);
        moveTo.setEndPosition(pose);
        moveTo.setMoveStraight(false);
        instance.setCRCLCommand(moveTo);
        s.writeCommand(instance, JNI_TRUE);

        BigInteger IDback= BigInteger::getONE();
        CommandStatusType cmdStat;

        do {
            GetStatusType getStat;
            getStat.setCommandID(BigInteger::valueOf(9));
            instance.setCRCLCommand(getStat);
            s.writeCommand(instance);

            CRCLStatusType stat = s.readStatus();
            cmdStat = stat.getCommandStatus();
            IDback = cmdStat.getCommandID();
            PrintObject("Command ID=", IDback);
            PrintObject("stat=", stat);
            PrintObject("cmdStat.getCommandState()=",cmdStat.getCommandState());
            pose = stat.getPoseStatus().getPose();
            PrintObject("pose=", pose);
            pt = pose.getPoint();
            PrintObject("X:", pt.getX());
            PrintObject("Y:", pt.getY());
            PrintObject("Z:", pt.getZ());
            JointStatusesType jst = stat.getJointStatuses();
            if (jst.jthis != NULL) {
                List l = jst.getJointStatus();
                for (int i = 0; i < l.size(); i++) {
                    JointStatusType elem;
                    elem = JointStatusType::cast(l.get(i));
                    PrintObject("Joint Number :", elem.getJointNumber());
                    PrintObject("Joint Position :", elem.getJointPosition());
                }
            }
        } while (!IDback.equals(moveTo.getCommandID()) || cmdStat.getCommandState().equals(CommandStateEnumType::getCRCL_WORKING()));
        cout << " End of C++ main() reached. " << endl;
    } catch (jthrowable t) {
        PrintJThrowable("Exception Thrown : ", t);
        exit(1);
    }
}